kotlin-codegen/runtime/src/main/java/com/squareup/moshi/MoshiSerializable.kt

```diff
@@ -0,0 +1,116 @@
+ */
+package com.squareup.moshi
+
+import java.lang.ref.WeakReference
+import java.lang.reflect.Constructor
+import java.lang.reflect.InvocationTargetException
+import java.lang.reflect.ParameterizedType
+import java.lang.reflect.Type
+import java.util.Collections
+import java.util.LinkedHashMap
+import kotlin.annotation.AnnotationRetention.RUNTIME
+import kotlin.annotation.AnnotationTarget.CLASS
+
+@Retention(RUNTIME)
+@Target(CLASS)
+annotation class MoshiSerializable
+
+class MoshiSerializableFactory : JsonAdapter.Factory {
+  private val adapters = Collections.synchronizedMap(
+      LinkedHashMap<Class<*>, Constructor<out JsonAdapter<*>>>())
+
+  override fun create(type: Type, annotations: MutableSet<out Annotation>,
+      moshi: Moshi): JsonAdapter<*>? {
+
+    val rawType = Types.getRawType(type)
+    if (!rawType.isAnnotationPresent(MoshiSerializable::class.java)) {
+      return null
+    }
+
+    val constructor = findConstructorForClass(rawType) ?: return null
+
+    try {
+      return if (constructor.parameterTypes.size == 1) {
+        constructor.newInstance(moshi)
+      } else {
+        if (type is ParameterizedType) {
+          constructor.newInstance(moshi, type.actualTypeArguments)
+        } else {
+          throw IllegalStateException("Unable to handle type $type")
+        }
+      }
+    } catch (e: IllegalAccessException) {
+      throw RuntimeException("Unable to invoke " + constructor, e)
+    } catch (e: InstantiationException) {
+      throw RuntimeException("Unable to invoke " + constructor, e)
+    } catch (e: InvocationTargetException) {
+      val cause = e.cause
+      if (cause is RuntimeException) {
+        throw cause
+      }
+      if (cause is Error) {
+        throw cause
+      }
+      throw RuntimeException(
+          "Could not create generated JsonAdapter instance for type " + rawType, cause)
+    }
+
+  }
+
+  private fun findConstructorForClass(cls: Class<*>): Constructor<out JsonAdapter<*>>? {
+    var adapterCtor: Constructor<out JsonAdapter<*>>? = adapters[cls]
+    if (adapterCtor != null) {
+      return adapterCtor
+    }
+    val clsName = cls.name.replace("$", "_")
+    if (clsName.startsWith("android.")
+        || clsName.startsWith("java.")
+        || clsName.startsWith("kotlin.")) {
+      return null
+    }
+    try {
+      val bindingClass = cls.classLoader
+          .loadClass(clsName + "_JsonAdapter")
+      adapterCtor = try {
+        // Try the moshi constructor
+        @Suppress("UNCHECKED_CAST")
+        bindingClass.getConstructor(
+            Moshi::class.java) as Constructor<out JsonAdapter<*>>
+      } catch (e: NoSuchMethodException) {
+        // Try the moshi + type constructor
+        @Suppress("UNCHECKED_CAST")
+        bindingClass.getConstructor(Moshi::class.java,
+            Array<Type>::class.java) as Constructor<out JsonAdapter<*>>
+      }
+
+    } catch (e: ClassNotFoundException) {
+      adapterCtor = findConstructorForClass(cls.superclass)
```

Comment:
I find this surprising. Presumably this doesn’t work for deserialization; it would return the wrong type

Reply:
I think this might be a vestige of the buterknife code this is adapted from. Agreed it shouldn't be necessary

Reply:
b9b2527